File: online_judges/kattis/bigtruck.cpp
#include <bits/stdc++.h>

using namespace std;
typedef pair<int,int> pi;

const int OO = (int)1e09;
typedef vector<pi> vpi;


struct place{
	int items,distance,node;
	place(int _items,int _distance,int _node){
		items = _items , distance = _distance, node = _node;
	}
	bool operator < (place p) const {
		if(items == p.items){
			return distance >= p.distance;
		}
		return items < p.items;
		
	}
};

int main(){
//	freopen("input","r",stdin);
	int n,m,a,b,d,item;
	scanf("%d",&n);
	int t[n+1];
	for(int i=1;i<=n;i++)
		scanf("%d",&t[i]);
	
	
	vector<vpi> g(n+1,vpi());
	scanf("%d",&m);
	for(int i=1;i<=m;i++){
		scanf("%d %d %d",&a,&b,&d);
		g[a].push_back(make_pair(b,d));
		g[b].push_back(make_pair(a,d));
	}
	vector<int> dis(n+1,OO);
	vector<int> items(n+1,0);
	

	priority_queue<place> q;
	dis[1] = 0;
	items[1] = t[1];

	q.push(place(items[1],0,1));
	while(!q.empty()){

		place p = q.top(); q.pop();

		item = p.items , a = p.distance , b = p.node;
		
		for(int j=0;j<g[b].size();j++){
			int e = g[b][j].first , w = g[b][j].second;
			if(dis[b] + w < dis[e] || (dis[b] + w == dis[e] && items[b] + t[e] > items[e])){
				dis[e] = dis[b] + w;
				items[e] = items[b] + t[e];
				q.push(place(items[e],dis[e],e));

			}
		}
	}
	if(dis[n] == OO) printf("impossible\n");
	else printf("%d %d",dis[n], items[n]);
	
	return 0;
}

File: online_judges/kattis/bigtruck.java

import java.io.*;
import java.util.*;


public class Solution {
    static int[] dx = {0, 1, 0, -1};
    static int[] dy = {1, 0, -1, 0};
    static int OO = (1 << 31) - 1;
    static List<Pair>[] graph;

    static char[][] grid;


    public static void main(String[] args){
        InputStream is;
        try {
            is = new FileInputStream(".\\src\\in.txt");
        } catch (FileNotFoundException e) {
            is = System.in;
        }
        InputReader in = new InputReader(is);
        PrintWriter out = new PrintWriter(System.out);

        int n = in.nextInt();
        graph = new List[n+1];
        for (int i=0;i<=n;i++) graph[i] = new ArrayList<>();

        int[] t = new int[n+1];
        for(int i=1;i<=n;i++) t[i] = in.nextInt();

        int m = in.nextInt();
        for(int i=0;i<m;i++){
            int u = in.nextInt();
            int v = in.nextInt();
            int w = in.nextInt();

            graph[u].add(new Pair(v, w));
            graph[v].add(new Pair(u, w));
        }

        int[] dist = new int[n+1];
        int[] items = new int[n+1];
        for (int i=0;i<=n;i++) {
            dist[i] = OO;
            items[i] = 0;
        }

        PriorityQueue<Place> pq = new PriorityQueue<>();
        dist[1] = 0;
        items[1] = t[1];

        pq.add(new Place(items[1],0,1));
        while (!pq.isEmpty()) {
            Place place = pq.poll();
            int item = place.items , a = place.distance , b = place.node;

            for(int j=0;j<graph[b].size();j++){
                Pair p = graph[b].get(j);
                int e = p.v , w = p.w;
                if(dist[b] + w < dist[e] || (dist[b] + w == dist[e] && items[b] + t[e] > items[e])){
                    dist[e] = dist[b] + w;
                    items[e] = items[b] + t[e];
                    pq.add(new Place(items[e], dist[e],e));

                }
            }

        }
        if(dist[n] == OO) out.println("impossible");
        else out.println(dist[n] + " " + items[n]);

        out.close();
    }

    static class Pair {
        public int v, w;

        public Pair(int v, int w) {
            this.v = v;
            this.w = w;
        }
    }

    static class Place implements Comparable{
        public int items,distance,node;


        public Place(int items, int distance, int node) {
            this.items = items;
            this.distance = distance;
            this.node = node;
        }

        @Override
        public int compareTo(Object o) {
            Place p = (Place)o;
            if (items == p.items) {
                if (distance < p.distance) return 1;
                if (distance == p.distance) return 0;
                return -1;
            }
            if (items < p.items) return -1;
            return 1;
        }
    }
    static class InputReader {
        public BufferedReader reader;
        public StringTokenizer tokenizer;

        public InputReader(InputStream stream) {
            reader = new BufferedReader(new InputStreamReader(stream), 32768);
            tokenizer = null;
        }

        public String next() {
            while (tokenizer == null || !tokenizer.hasMoreTokens()) {
                try {
                    tokenizer = new StringTokenizer(reader.readLine());
                } catch (IOException e) {
                    throw new RuntimeException(e);
                }
            }
            return tokenizer.nextToken();
        }

        public int nextInt() {
            return Integer.parseInt(next());
        }
        public long nextLong() {
            return Long.parseLong(next());
        }
        public double nextDouble() {return Double.parseDouble(next());}
    }
}